File: c/fizzbuzz.c
#include <stdbool.h>
#include <stdio.h>

#define FIZZ "Fizz"
#define BUZZ "Buzz"

static bool fizz(int n) { 
    return n % 3 == 0;
}

static bool buzz(int n) {
    return n % 5 == 0;
}

int main(void) {
    for(int i = 1; i <= 100; ++i) {
        if(fizz(i) && buzz(i)) {
            printf(FIZZ BUZZ "\n");
        } else if(fizz(i)) {
            printf(FIZZ "\n");
        } else if(buzz(i)) {
            printf(BUZZ "\n");
        } else {
            printf("%d\n", i);
        }
    }
}

File: perl/fizzbuzz.pl
#!/usr/bin/env perl

use strict;
use warnings;
use feature qw/say/;

sub get_or_empty {
    my $text = shift;
    my $test = \&{shift @_};
    return $test->() ? $text : '';
}

sub default_if_empty {
    my $textref = shift;
    my $default = shift;
    return length($$textref) ? $$textref : $default;
}

for (1..100) {
    my $fizz = get_or_empty fizz => sub { $_ % 3 == 0 };
    my $buzz = get_or_empty buzz => sub { $_ % 5 == 0 };
    my $fizzbuzz = bless \"$fizz$buzz";
    say $fizzbuzz->default_if_empty($_);
}
